### PR TITLE
Improvements to RDNodeResource

### PR DESCRIPTION
**Pull request by navoda**
The method `addNodeResource` changed to `addResource` making it more generalized when adding new resources with attributes.
Added `addEndResource` method to modifying purposes, like dynamically adding resources with specific resource types according to receiving attributes (e.g. rt, if)
also added a `lifeTime` getter and changed `updateEndpointResources `'s access modifier from private to public.

Reference to the [Improvement to RDNodeResource #26](https://github.com/eclipse/californium.tools/pull/26)
Signed-off-by: navoda <mpnavoda@gmail.com>

### CODE REVIEW

File: cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java [hunk 1]
```diff
@@ -49,7 +50,7 @@
 	 * if it does not respond.
 	 */
 	private static ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(//
-			new Utils.DaemonThreadFactory("RDLifeTime#"));
+			new Utils.DaemonThreadFactory());
```
Comment (sophokles73) [inline cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java]: There is no default constructor defined on Californium master. We need to keep it like it is ...

File: cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java [hunk 2]
```diff
@@ -207,7 +211,21 @@ public CoapResource addNodeResource(String path) {
...
+	/**
+     * this adds the end resource of a path with respect to the resource attributes
+     * eg: In the /sensors/door path this method adds the 'door' resource with it's resource attributes ct=41;rt="door";if="sensor"
+     *
+     * @param name       - name of the resource adding
+     * @param attributes - the specific attributes the resource has [ct, rt, if]
+     * @return 	- the resource with changed resource type
+     * @override - resource can be set to a sensor, actuator, composite or any other interface description
+     */
+	public CoapResource addEndResource(String name, ResourceAttributes attributes) {
```
Comment (sophokles73) [inline cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java]: why is the `attributes` param defined if it is not used in the method body?

File: cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java [hunk 3]
```diff
@@ -179,7 +180,7 @@ private void setContextFromRequest(Request request, String newContext)
 	 * humidity. If the path is /readings/temp, temp will be a subResource
 	 * of readings, which is a subResource of the node.
 	 */
-	public CoapResource addNodeResource(String path) {
+	public CoapResource addResource(String path, ResourceAttributes attributes) {
```
Comment (sophokles73) [inline cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java]: I do not understand why you want to pass in the attributes here? you can simply set the attributes on the returned `CoapResource`:
````
CoapResouce newRes = addNodeResouce("my/path");
newRes.getAttributes().addResourceType("myType");
...
````

File: cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java [hunk 4]
```diff
@@ -197,7 +198,10 @@ public CoapResource addNodeResource(String path) {
 				}
 			}
 			if (!resourceExist) {
-				subResource = new RDTagResource(next, true, this);
+				if (scanner.hasNext())//if the resource is not the end resource
+					subResource = new RDTagResource(next, true, this);
+				else
+					subResource = addEndResource(next, attributes);
```
Comment (sophokles73) [inline cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java]: what's the purpose of this if the `addEndResource()` method also adds an `RDTagResource`?

File: cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java [hunk 5]
```diff
@@ -289,22 +307,22 @@ public void run() {
 	 * register a resource for reading the temperature and another one
 	 * for reading the humidity.
 	 */
-	private boolean updateEndpointResources(String linkFormat) {
+	public boolean updateEndpointResources(String linkFormat) {
```
Comment (sophokles73) [inline cf-rd/src/main/java/org/eclipse/californium/tools/resources/RDNodeResource.java]: why do you need to make this public?

**Comment by boaks** [comment]
I'm not sure, if you still work on this. I guess no.

**Pull request closed without merging**